File: 461. Hamming Distance.cpp
class Solution {
public:
    // bitset: O(1)time & space, 1.95%, ok
    int hammingDistance(int x, int y) {
        bitset<32> x1(x);
        bitset<32> y1(y);
        int ret = 0;
        for(int i = 0; i < 32; i++)
            ret += x1[i] != y1[i];
        return ret;
    }

    // bit: O(1)time & space, 1.95%, ok
    int hammingDistance(int x, int y) {
        int ret = 0;
        for(int i = 0; i < 32; i++){
            ret += (x & 1) != (y & 1);
            x = x >> 1;
            y = y >> 1;
        }
        return ret;
    }

    // exclusive or: O(1)time & space, 1.95%, ok
    int hammingDistance(int x, int y) {
        bitset<32> ret(x ^ y);
        return ret.count();
    }

    // erase rightmost 1 bit: O(1)time & space, 1.95%, ok
    int hammingDistance(int x, int y) {
        int ret = 0, n = x ^ y;
        while(n){
            ret++;
            n &= n - 1;
        }
        return ret;
    }
};

File: 461. Hamming Distance.py
class Solution(object):
    
    # Bit manipulation: O(N) time, O(1) space, 77.90%
    def hammingDistance(self, x, y):
        """
        :type x: int
        :type y: int
        :rtype: int
        """
        dist = 0
        for i in xrange(32):
            dist += ((1 << i) & x > 0) != ((1 << i) & y > 0)
        return dist
        
    # Bit manipulation: O(N) time, O(1) space, 59.69%
    def hammingDistance(self, x, y):
        xor, dist = x ^ y, 0
        for i in xrange(32):
            dist += ((1 << i) & xor > 0)
        return dist
    
    # Bit manipulation: O(N) time, O(1) space, 18.75%
    def hammingDistance(self, x, y):
        return bin(x ^ y).count('1')